File: nodejs.py
import json
import os
from common import Storage

storage = Storage('nodejs')


def is_compress_file(__path__):
    return '.tar.gz' in __path__ \
        or '.tar.xz' in __path__ \
        or '.zip' in __path__ \
        or '.7z' in __path__


def is_ignored_path(__path__):
    return 'latest' in __path__ \
        or 'npm' in __path__ \
        or 'patch' in __path__ \
        or '-isaacs-manual' in __path__


def exec_cmd_json(__cmd__):
    return json.loads(os.popen(__cmd__).read())


def main():
    remote_site = 'https://nodejs.org/dist/'
    versions = exec_cmd_json('rclone lsjson --http-url {} :http:'.format(remote_site))
    idx = 0
    for version in versions:
        idx += 1
        print('{} / {}    {}'.format(idx, len(versions), version['Path']))
        if not version['IsDir'] or is_ignored_path(version['Path']):
            continue
        try:
            items = exec_cmd_json('rclone lsjson --http-url {}{}/ :http:'.format(remote_site, version['Path']))
            for v in items:
                if not is_compress_file(v['Path']):
                    continue
                path = '{}/{}'.format(version['Path'], v['Path'])
                origin_url = '{}{}/{}'.format(remote_site, version['Path'], v['Path'])
                storage.write(path, origin_url)
        except Exception as e:
            print('Err: {}'.format(e))


if __name__ == '__main__':
    main()

File: task/nodejs.go
package task

import (
	"fmt"
	"strings"

	"github.com/ismdeep/mirror-data/conf"
	"github.com/ismdeep/mirror-data/global"
	"github.com/ismdeep/mirror-data/internal/rclone"
	"github.com/ismdeep/mirror-data/internal/store"
	"github.com/ismdeep/mirror-data/internal/util"
	"github.com/ismdeep/mirror-data/pkg/log"
	"go.uber.org/zap"
)

type NodeJS struct {
}

func (receiver *NodeJS) Run() {
	storage := store.New("nodejs", conf.Config.StorageCoroutineSize)
	defer func() {
		storage.CloseAndWait()
	}()

	versions, err := rclone.JSON("lsjson", "--http-url", "https://nodejs.org/dist/", ":http:")
	if err != nil {
		global.Errors <- err
		return
	}

	versionChan := make(chan rclone.JSONObj, 1024)
	go func() {
		for _, version := range versions {
			versionChan <- version
		}
		close(versionChan)
	}()

	err = util.CoroutineRun(conf.Config.CoroutineSize, func() error {
		for version := range versionChan {
			if !version.IsDir || receiver.isIgnoredPath(version.Path) {
				continue
			}
			items, err := rclone.JSON("lsjson", "--http-url", fmt.Sprintf("https://nodejs.org/dist/%v/", version.Path), ":http:")
			if err != nil {
				log.WithName("nodejs").Error("failed on lsjson", zap.Error(err))
				return err
			}
			for _, v := range items {
				if !receiver.isCompressFile(v.Path) {
					continue
				}
				link := fmt.Sprintf("%v/%v", version.Path, v.Path)
				originLink := fmt.Sprintf("https://nodejs.org/dist/%v/%v", version.Path, v.Path)
				storage.Add(link, originLink)
			}
		}
		return nil
	})
	if err != nil {
		global.Errors <- err
		return
	}
}

// isCompressFile 是否是压缩包
func (receiver *NodeJS) isCompressFile(path string) bool {
	return strings.Contains(path, ".tar.gz") ||
		strings.Contains(path, ".tar.xz") ||
		strings.Contains(path, ".zip") ||
		strings.Contains(path, ".7z")
}

// isIgnoredPath 是否被忽略
func (receiver *NodeJS) isIgnoredPath(path string) bool {
	return strings.Contains(path, "latest") ||
		strings.Contains(path, "npm") ||
		strings.Contains(path, "patch") ||
		strings.Contains(path, "-isaacs-manual")
}
